### ISSUE TITLE
Authoring flow segments and fittings

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#2181]
The goal is build operators that can perform connections between flow segments, and automatically insert flow fittings.

IFC does not establish any conventions around axis orientations for segments and fittings. I had a look at what Revit does, but their system is pretty unsemantic, where all object origins are at the storey, and e.g. for a rectangular ducts their rectangle X and Y parameters don't always correlate to the cross section. It's pretty terrible. So I'm going to follow what IFC does for beams / columns / profile sets which makes a ton more sense.

When creating a rectangular duct segment, the axis convention must always be like this: the duct cross section width is local X, height is local Y, and it is always extruded in the positive Z direction. The position must always be in the center as shown. This matches the convention for IFC profiles exactly.

![2022-05-03-121403_473x393_scrot](https://user-images.githubusercontent.com/88302/166395429-5b03cbe7-8870-40ed-8e12-dae6c2741202.png)

![image](https://user-images.githubusercontent.com/88302/166395509-f168d871-02cc-492d-b713-7609ce679c92.png)

The user must be able to choose a point with the 3D cursor, and use Shift-E to "extend" the point closest to the cursor. Similar to how extending walls works.

The user must also be able to choose a point with the 3D cursor and use Shift-S to "split" at the point closest to the cursor. Similar to how splitting walls works.

In addition, the following operations need to be possible:

 - Bend
 - Connector
 - Junction (more than 2 connections)
 - Obstruction (only needs one segment as input)
 - Transition (same as connector, but occurs between segments of two different types)

When a connection between two segments occurs a third object will be inserted called the fitting. The fitting geometry will be already available, so the only requirement is to derive the matrix_world of the fitting.

See screenshots of typical operations:

[mep-axis.blend.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/8608065/mep-axis.blend.txt)

![2022-05-03-124438_761x739_scrot](https://user-images.githubusercontent.com/88302/166399362-f06e07ac-b048-42a5-87d4-ee7f5a9968de.png)

**Comment by aothms** [comment]
+1 on extrude towards +Z

Sent from a mobile device. Excuse my brevity. Kind regards, Thomas

Op di 3 mei 2022 04:53 schreef Dion Moult ***@***.***>:

**Comment by Moult** [comment]
@aothms do you know whether or not there is an IFC relationships available to describes that an IfcFlowFitting**Type** is capable of connecting one or more other IfcFlowSegmentTypes?

Unless I missed something, there seems to be no way of knowing whether or not a flow fitting type is applicable to a segment type.

The PredefinedType seems to help narrow the scenarios (e.g. a BEND is only applicable to 2 segments sharing the same type where as a TRANSITION is applicable to two segments of different types) The JunctionType in Pset_FittingJunction can also help filter 3 and 4 segment combos, but not exactly which segment type it applies to. And of course the class (e.g. duct fittings only connect ducts, not pipes) also helps.

If nothing is available maybe IfcRelAssignsToProduct can be (ab)used for this purpose?

**Comment by aothms** [comment]
> @aothms do you know whether or not there is an IFC relationships available to describes that an IfcFlowFittingType is capable of connecting one or more other IfcFlowSegmentTypes?

No, sorry, I'm very unfamiliar with the semantic structure behind most of the systems in IFC

**Comment by Moult** [comment]
OK, for the moment then as an authoring tool I can use the existing hints to narrow down a shortlist, then the user can pick from there. Once a fitting is used, however, it then becomes very easy to know which one its applicable to... just check where it's already used :) Not ideal of course.

**Comment by Isaina** [comment]
Here is a sample Acca is reference in IFc world also member of ODA so is reference point for MEP authoring tool but is close source and can’t write plug-ins for academic world

https://youtu.be/bZHbG_BhzcU?feature=shared

**Comment by KakaoFelix** [comment]
I think i understood 75% of it :D 
But my impression was aswell, if you have the dimensions and the volume you can let the program calcuate it and give you the speed. 
Would be good if there was no hard cap on 4m/s because you can go over that in certain areas. When there is to much installations you sometimes need to go over 4m/s otherwise you have to many collisions with walls or other pipes etc.

I cant really help with any technical solutions, only wanted to share what i think would be cool to have in such a tool, or rather what i could use, since thats what i need for my work.

**Comment by Isaina** [comment]
Basically you need to have volume flow defined in all terminal discharge points, but for that in each room, space you need to define ventilation method, one is AHI basically take room volume (area x height) and multiple with 6-8 1/h ( but can be one number from 0-20 1/h this method are for toilets, baths, restaurants etc., A second is define amount of person in room (number of person x 35m3/h/person) and with this you will get needed air volume for each room space ( m3/h ). Now you must check is sum of all air terminal in room equal or little bigger of flow needed for this room. Basically you must have small parametric database of air terminal with size and nominal size like (round air valve , fi 100, 90 m3/h) , (round air valve , fi125, 120 m3/h), and for grill size with plenums.

**Comment by Isaina** [comment]
here is a good article for that

https://www.tandfonline.com/doi/pdf/10.1080/13467581.2022.2077734

**Comment by Moult** [comment]
This is awesome! We're not yet at the stage of running calculations - still implementing geometric authoring and port connectivity. After that, definitely the next step is calculations!

**Comment by Isaina** [comment]
One more idea , is possible to make like native ifc , git respiratory for hvac elements, furniture so we can make ifc elements as contributor , hvac element maybe parametric like 

https://help.graphisoft.com/AC/24/int/_AC24_Help/130_UserInterfaceDialogBoxes/130_UserInterfaceDialogBoxes-65.htm
